File: frontend/src/api/brand.ts
/** Brand settings endpoints — see docs/brand-api-contract.md. */
import type { DefaultMode, PartialBrandTokens } from '@/theme/tokens';
import { api, request } from './client';

export interface BrandRecord {
  id: string;
  name: string;
  /** Applied preset code, or `custom`. */
  preset: string;
  /** Stored token override (deep-partial, merged over the platform defaults). */
  tokens: PartialBrandTokens;
  updated_at: string;
}

export interface BrandPreset {
  code: string;
  name: string;
  description?: string;
  /** Colors shown on the selection tile. */
  swatch: string[];
  default_mode?: DefaultMode;
  /** Full token set the preset applies. */
  tokens: PartialBrandTokens;
}

export interface BrandAbstraction {
  code: string;
  name: string;
  preview_url: string;
}

export interface BrandFont {
  /** Exact string that goes into `typography.fontFamily`. */
  family: string;
  name: string;
  category: string;
}

export function fetchBrand(): Promise<BrandRecord> {
  return api.get<BrandRecord>('/cms/brand');
}

/** Partial token patch — deep-merged on the server over the current set. */
export function patchBrand(tokens: PartialBrandTokens): Promise<BrandRecord> {
  return api.patch<BrandRecord>('/cms/brand', { tokens });
}

/**
 * Данные одного экрана показа — в той же форме, в какой их получает гость.
 *
 * Ходит в ПАНЕЛЬНУЮ ручку, а не в гостевую: гостевая требует сессию, и заводить
 * её на каждый заход в оформление значило бы плодить гостей, которых не было.
 * Считают ответ те же сборщики, что отвечают витрине.
 */
export function fetchPreviewScreen(
  screen: string,
  params: Record<string, string> = {},
): Promise<unknown> {
  return api.get<unknown>('/cms/brand/preview', { query: { screen, ...params } });
}

export function fetchPresets(): Promise<{ presets: BrandPreset[] }> {
  return api.get<{ presets: BrandPreset[] }>('/cms/brand/presets');
}

export function applyPreset(preset: string): Promise<BrandRecord> {
  return api.post<BrandRecord>('/cms/brand/apply-preset', { preset });
}

export function fetchAbstractions(): Promise<{ abstractions: BrandAbstraction[] }> {
  return api.get<{ abstractions: BrandAbstraction[] }>('/cms/brand/abstractions');
}

export function fetchFonts(): Promise<{ fonts: BrandFont[] }> {
  return api.get<{ fonts: BrandFont[] }>('/cms/brand/fonts');
}

export interface UploadedFont {
  assetId: string;
  name: string;
  family: string;
  url: string;
  format: string;
}

/**
 * Загрузка файла шрифта отеля.
 *
 * Токены НЕ трогает: загрузить и выбрать — разные решения оператора. Ответ
 * несёт готовую строку семейства, а положит её в тему редактор — когда
 * оператор этого захочет.
 */
export function uploadBrandFont(file: File): Promise<UploadedFont> {
  const form = new FormData();
  form.append('file', file);
  return request<UploadedFont>('/cms/brand/font', { method: 'POST', formData: form });
}

File: backend/api/cms/brand.py
"""
CMS: бренд-настройки. Контракт — docs/brand-api-contract.md.

Вьюхи тонкие: вся валидация и merge — в apps/hotels/brand_services.py.
"""

from __future__ import annotations

from typing import Any

from django.http import HttpRequest
from ninja import Router, Schema

from apps.hotels import brand_services as svc
from apps.hotels.brand_library import ABSTRACTIONS, FONTS, list_presets

router = Router(tags=["cms:brand"])


class BrandOut(Schema):
    id: str
    name: str
    preset: str
    tokens: dict[str, Any]
    updated_at: str


class BrandPatch(Schema):
    tokens: dict[str, Any] = {}


class ApplyPresetIn(Schema):
    preset: str


@router.get("/brand", response=BrandOut, summary="Текущая тема отеля")
def get_brand(request: HttpRequest):
    return svc.serialize_brand(svc.get_or_create_brand())


@router.patch("/brand", response=BrandOut, summary="Изменить токены (deep-merge)")
def patch_brand(request: HttpRequest, payload: BrandPatch):
    theme = svc.update_brand(payload.tokens or {})
    return svc.serialize_brand(theme)


@router.get("/brand/presets", summary="Библиотека пресетов")
def presets(request: HttpRequest):
    return {"presets": list_presets()}


@router.post("/brand/apply-preset", response=BrandOut, summary="Применить пресет целиком")
def apply_preset(request: HttpRequest, payload: ApplyPresetIn):
    theme = svc.apply_preset(payload.preset)
    return svc.serialize_brand(theme)


@router.get("/brand/fonts", summary="Курируемый список шрифтов")
def fonts(request: HttpRequest):
    return {"fonts": FONTS}


@router.get("/brand/abstractions", summary="Библиотека фонов-абстракций")
def abstractions(request: HttpRequest):
    return {"abstractions": ABSTRACTIONS}
